larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc

```diff
@@ -0,0 +1,451 @@
+        {
+            if ((*iIter) == eIter2)
+                continue;
+
+            unsigned int nMatchedClusters(0);
+
+            if ((*iIter)->GetCluster1() == eIter2->GetCluster1())
+                ++nMatchedClusters;
+
+            if ((*iIter)->GetCluster2() == eIter2->GetCluster2())
+                ++nMatchedClusters;
+
+            if (nMatchedClusters > 1)
+                continue;
+
+            if (nMatchedClusters)
+            {
+                iteratorList.push_back(eIter2);
+                return;
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TwoViewThreeDKinkTool::GetIteratorListModifications(
+    TwoViewTransverseTracksAlgorithm *const pAlgorithm, const IteratorList &iteratorList, ModificationList &modificationList) const
+{
+    for (IteratorList::const_iterator iIter1 = iteratorList.begin(), iIter1End = iteratorList.end(); iIter1 != iIter1End; ++iIter1)
+    {
+        for (IteratorList::const_iterator iIter2 = std::next(iIter1); iIter2 != iIter1End; ++iIter2)
+        {
+            try
+            {
+                const unsigned int nMatchedReUpsampledSamplingPoints1((*iIter1)->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints());
+                const unsigned int nMatchedReUpsampledSamplingPoints2((*iIter2)->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints());
+                IteratorList::const_iterator iIterA((nMatchedReUpsampledSamplingPoints1 >= nMatchedReUpsampledSamplingPoints2) ? iIter1 : iIter2);
+                IteratorList::const_iterator iIterB((nMatchedReUpsampledSamplingPoints1 >= nMatchedReUpsampledSamplingPoints2) ? iIter2 : iIter1);
+
+                Particle particle(*(*iIterA), *(*iIterB));
+                const LArPointingCluster pointingClusterA(pAlgorithm->GetCachedSlidingFitResult(particle.m_pClusterA));
+                const LArPointingCluster pointingClusterB(pAlgorithm->GetCachedSlidingFitResult(particle.m_pClusterB));
+
+                LArPointingCluster::Vertex vertexA, vertexB;
+                LArPointingClusterHelper::GetClosestVertices(pointingClusterA, pointingClusterB, vertexA, vertexB);
+
+                float transverseAB(std::numeric_limits<float>::max()), transverseBA(std::numeric_limits<float>::max());
+                float longitudinalAB(-std::numeric_limits<float>::max()), longitudinalBA(-std::numeric_limits<float>::max());
+
+                LArPointingClusterHelper::GetImpactParameters(vertexA, vertexB, longitudinalAB, transverseAB);
+                LArPointingClusterHelper::GetImpactParameters(vertexB, vertexA, longitudinalBA, transverseBA);
+
+                if (std::min(longitudinalAB, longitudinalBA) < m_minLongitudinalImpactParameter)
+                    continue;
+
+                if (std::min(transverseAB, transverseBA) > m_maxTransverseImpactParameter)
```

Comment:
Should this one be `std::max(args)` ?

Reply:
So, here as well I followed the same philosophy of an existing tool (the UndershootTracksTool, [here](https://github.com/PandoraPFA/LArContent/blob/master/larpandoracontent/LArThreeDReco/LArTransverseTrackMatching/UndershootTracksTool.cc#L68)), but I do see your point that max is a stronger condition and more appropriate. I don't know if the original one was a choice or a bug!

Reply:
Yes you're right - the logic here also comes from an existing tool (the UndershootTracksTool [here](So, here as well I followed the same logic of an existing tool (the UndershootTracksTool, here), but I do see your point that max is probably more appropriate. I don't know if the original one was a choice or a bug!)) but as you point out I think it was probably a bug there?

Reply:
(I managed to write 3 copies of this comment before it came out as I intended it to... :D)

Reply:
Hi, I think this is correct as is, and correct in the original tools. It's taking the smallest available impact parameter, and then checking to see if that value is still bigger than some threshold check. So pretty sure we want to keep this intact (unless want to start up some explicit study to check).

Reply:
Does this not allow for the case when the minimum is smaller than m_maxTransverseImpactParameter, but the maximum isn't? (It may be that things are getting confused in my head)

Reply:
It's been a while now(!), but I think it's looking at the consistency of vertex A and B in two ways. One using a position extrapolated to B, based on the position and direction estimated at A; the other using a position extrapolated to A, based on the position and direction estimated at B. Good agreement (small values) either way around would indicate some kind of association. So, we take the smallest values (representing the most plausible agreement, we think) and then we apply quality thresholds to those values. Would need to start drawing some diagrams out, and thinking through the surrounding logic too, to completely refamiliarise with the issues at stake. Hope that helps!

Reply:
I see, thanks John - I thought this condition was meant to ensure that both transverseAB and transverseBA were below the m_maxTransverseImpactParameter threshold, just like the condition above ensures that longitudinalAB and longitudinalBA are larger than m_minLongitudinalImpactParameter. So I was under the wrong assumption on the intention behind the checks.

Reply:
Yes, sometimes the only thing you can do here is draw out an example event, think through the logic you want and check that the code is in agreement. Here, can imagine a 2D kink topology, with two tracks in something like a v-shape, but a bit of a gap between the track endpoints (i.e. a chunk missing from one of the tracks).

Extrapolating one of the tracks will successfully allow you to reach the vertex of the second track, but extrapolating the other way around will not do the same thing. We use the successful extrapolation here to say that the tracks might be associated, but then apply some quality cuts to that possible association.

Reply:
Yes, I see, that makes sense! Thanks for the clarifications!

Reply:
I am little confused here.
With the way this is configured now, only one of the A,B clusters has to look like it passes in close proximity to the other cluster?  Do you not need the double coincidence to know that the two clusters actually do align?